v3 with asyncio

furkansimsekli opened the pull request:
Finally I'm opening this PR!

## PTBv20
There have been lots of changes. The wrapper we use, PTB, changed a lot in v20. One of the biggest change was switching to asyncio from multithreading. That's why Hacettepe Duyurucusu became asyncio as well. Transition to asyncio required me to change libs like [requests](https://requests.readthedocs.io/en/latest/) and [pymongo](https://pymongo.readthedocs.io/en/stable/), therefore requests replaced by [aiohttp](https://docs.aiohttp.org/en/stable/) and pymongo replaced by [motor](https://motor.readthedocs.io/en/stable/). For those who wonders why this transition has been made, it's because requests & pymongo aren't built on top of the asyncio like aiohttp & motor.

There are other minor changes due to v20, although I don't think they are worth to mention here. See [PTBv20 Changelog](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Transition-guide-to-Version-20.0).

## Scrapers
Unlike old version of HD, scraper codebase are much more clear in v3. They have been moved to `scraper` module and a minor inheritance has been implemented in order to reduce the number of methods.

## Handlers & Mongo
All handlers are also gathered in one place in order to reduce the complexity. This change allowed me to remove dummy modules like `Keyboard` and `Text`.

In the old version, I was using another `User` and `Announcement` module to abstract the design. Although, I noticed that this was an unnecessary work and made the things more bloated, therefore I removed them.

## Speed
There have been improvements at the performance side. Number of database requests have been **reduced more than %50**. Also translation became faster because in the old version of the HD was reading from the disk, however now it uses memory. I'm not sure if it's a good idea to fulfill memory with such purpose. If it becomes necessary, we may convert it back to old way. Remember that caching is an option too. Right now HD doesn't use any caching explicitly, in the future versions I'm thinking that we should implement a caching for database requests and translations!

## Logging
Logging is another thing that has been improved. As mentioned in #23, developers are being directly contacted on Telegram in case of errors in the system.

## 16 New Departments
- Department of Electrical and Electronics Engineering
- Physics Engineering
- Geomatics Engineering Department
- Department of Food Engineering
- Department of Mining Engineering
- Nuclear Engineering Department
- Mathematics Teaching Programme
- Elementary Mathematics Teaching Programme
- Department of Nutrition and Dietetics
- Faculty of Nursing
- European Union Office
- Disabled Student Unit
- Vocational School of Health Services
- Vocational School of Social Sciences
- Vocational School of Technical Sciences
- Elective Courses Coordinator

## Other
- Build became easier thanks to Makefile.
- Translation keys have become more understandable and translations themselves have been improved.
- Some configuration constants have been moved into config.
- Documentations have been improved.
- Resolved #19 and #23.
- Overall code standard has been improved.

## What else is coming??
- Friendlier messages with emojis 👀
- New bot commands for admins and standard users (For example, showing some stats would be cool)
- Student club integration. They will be able to publish their announcements from HD too!
- r/Hacettepe integration with Reddit API
- Instagram integration for popular Hacettepe related pages with Instagram API
- GitHub workflow!
- More departments...

furkansimsekli committed:
gitignore: use Python template

furkansimsekli committed:
backend: rewrite

Details: Lots of thins have changed in order to adapt the new version
of the PTB v20, but also to improve the performance and codebase.
More details will be on the PR.

furkansimsekli committed:
build: add Makefile

furkansimsekli committed:
app: add new departments

Details:
- Department of Electrical and Electronics Engineering (custom scraper)
- Physics Engineering (custom scraper)
- Geomatics Engineering Department
- Department of Food Engineering
- Department of Mining Engineering
- Nuclear Engineering Department
- Mathematics Teaching Programme
- Elementary Mathematics Teaching Programme
- Department of Nutrition and Dietetics
- Faculty of Nursing

furkansimsekli committed:
fix: await context.bot.send_message()

furkansimsekli committed:
docs: simplify and update according to changes

Details: Repository has been transferred into hacettepeoyt, therefore
urls have been changed. Also, new scraper guide has been written.

furkansimsekli committed:
locale: sort names by given char set

Details: As mentioned in #19, the names were being sorted by ASCII.
From now on, they will be sorted by the given character set.

melikechan commented:
Long awaited update... Seems awesome and structing why the changes have made also dissolved my confusions. :3

Let me know if you need any support for anything!

furkansimsekli committed:
fix: add control flow for span tag into ie scraper

furkansimsekli committed:
app: add new departments

Details:
- European Union Office
- Disabled Student Unit
- Vocational School of Health Services
- Vocational School of Social Sciences
- Vocational School of Technical Sciences
- Elective Courses Coordinator
